lib/hopper.rb

```diff
@@ -61,8 +88,32 @@ def clear
       @registrations = []
     end
 
+    def semaphore
+      @semaphore ||= Mutex.new
```

Comment:
I could be overthinking this, and Ruby's GIL might make this a nonissue: if you're calling this in a multithreaded scenario, multiple threads might read `@semaphore` and return `nil`, causing multiple new Mutex creations. It's conceivable that it might overwrite a mutex that is currently locked.

Reply:
@bvrooman 

What would be the solution. Is something like this ok:

```
  class << self
   attr_accessor :semaphore
  
  ...
  end

  Hopper.semaphore ||= Mutex.new
```

Reply:
`Hopper.semaphore = Mutex.new` would probably suffice since class loading is single threaded. But again, this may not even be an issue. In my experience, I have found that Ruby somehow (probably the GIL) avoids a lot of concurrent read/writes anyway. I'd say it's your call. We could probably wait to see if it ever is a problem